server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java

```diff
@@ -0,0 +1,164 @@
+     *
+     * @param path the path to the file
+     * @return an {@link OptionalLong} that contains the number of allocated bytes on disk for the file, or empty if the size is invalid
+     */
+    @Override
+    public OptionalLong allocatedSizeInBytes(Path path) {
+        assert Files.isRegularFile(path) : path;
+        try {
+            final Stat stats = new Stat();
+            final int rc = XStatLibrary.__xstat(STAT_VER_KERNEL, path.toString(), stats);
+            if (logger.isTraceEnabled()) {
+                logger.trace("executing native method __xstat() returned {} with error code [{}] for file [{}]", rc, stats, path);
+            }
+            return OptionalLong.of(stats.st_blocks * ST_BLOCKS_UNIT);
+        } catch (LastErrorException e) {
+            logger.warn(
+                () -> new ParameterizedMessage(
+                    "error when executing native method __xstat(int vers, const char *name, struct stat *buf) for file [{}]",
+                    path
+                ),
+                e
+            );
+        }
+        return OptionalLong.empty();
+    }
+
+    @Structure.FieldOrder(
+        {
+            "st_dev",
+            "st_ino",
+            "st_mode",
+            "st_nlink",
+            "st_uid",
+            "st_gid",
+            "st_rdev",
+            "st_size",
+            "st_blksize",
+            "st_blocks",
+            "st_atim",
+            "st_mtim",
+            "st_ctim",
+            "st_atim_tv_sec",
+            "st_mtim_tv_sec",
+            "st_ctim_tv_sec" }
+    )
+    public static class Stat extends Structure {
+
+        public long st_dev;         // __dev_t st_dev; /* Device. */
+        public long st_ino;         // __ino_t st_ino; /* File serial number. */
+        public int st_mode;         // __mode_t st_mode; /* File mode. */
+        public long st_nlink;       // __nlink_t st_nlink; /* Link count. */
+        public int st_uid;          // __uid_t st_uid; /* User ID of the file's owner. */
+        public int st_gid;          // __gid_t st_gid; /* Group ID of the file's group. */
+        public long st_rdev;        // __dev_t st_rdev; /* Device number, if device. */
+        public long st_size;        // __off_t st_size; /* Size of file, in bytes. */
+        public long st_blksize;     // __blksize_t st_blksize; /* Optimal block size for I/O. */
+        public long st_blocks;      // __blkcnt_t st_blocks; /* Number 512-byte blocks allocated. */
+        public Time st_atim;        // struct timespec st_atim; /* Time of last access. */
+        public Time st_mtim;        // struct timespec st_mtim; /* Time of last modification. */
+        public Time st_ctim;        // struct timespec st_ctim; /* Time of last status change. */
+        public long st_atim_tv_sec; // backward compatibility
+        public long st_mtim_tv_sec; // backward compatibility
+        public long st_ctim_tv_sec; // backward compatibility
```

Comment:
I suspect this puts `st_blocks` in the right place but it doesn't match the `struct stat` that my system sees:

```
$ uname -a
Linux worker-772973 4.15.0-120-generic #122~16.04.1-Ubuntu SMP Wed Sep 30 15:09:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ ldd --version
ldd (Ubuntu GLIBC 2.23-0ubuntu11.3) 2.23
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
$ echo "#include <sys/stat.h>" | gcc -xc - -E -dD | grep -ve '^$' | grep -A23 '^struct stat'
struct stat
  {
    __dev_t st_dev;
    __ino_t st_ino;
    __nlink_t st_nlink;
    __mode_t st_mode;
    __uid_t st_uid;
    __gid_t st_gid;
    int __pad0;
    __dev_t st_rdev;
    __off_t st_size;
    __blksize_t st_blksize;
    __blkcnt_t st_blocks;
# 91 "/usr/include/x86_64-linux-gnu/bits/stat.h" 3 4
    struct timespec st_atim;
    struct timespec st_mtim;
    struct timespec st_ctim;
#define st_atime st_atim.tv_sec
#define st_mtime st_mtim.tv_sec
#define st_ctime st_ctim.tv_sec
# 106 "/usr/include/x86_64-linux-gnu/bits/stat.h" 3 4
    __syscall_slong_t __glibc_reserved[3];
# 115 "/usr/include/x86_64-linux-gnu/bits/stat.h" 3 4
  };
```

In particular, `st_mode` and `st_nlink` are the other way around, there's 4 bytes of padding between `st_gid` and `st_rdev`, and the last three `long` fields are apparently unrelated to time.

Is this how some other `x86_64 GNU/Linux` system declares its `struct stat`? If so, I think we need to understand the differences.

(might also be worth recording the command `echo "#include <sys/stat.h>" | gcc -xc - -E -dD | grep -ve '^$' | grep -A23 '^struct stat'` in a comment so that future readers can reproduce this analysis)

Reply:
> 

Yes, I'm just trying to make sense of all of this, given the number of indirection each system can have given the compiling options that were provided.

Reply:
Ok, I think I finally managed to link the struct stat compiled on my system (and yours in a different GLIBC version) with the glibc source code. The way the glibc source includes header files depending of the platform is puzzling and there was more recent changes that confused me even more.

I updated the code to finally match what is expected by our systems. I also add documentation to link to the command you provided as well as the source files for the two glibc versions we analyzed.

Hopefully we should be in a better place now 🤞🏻